add lttf process blog post (#2633)
mainmatter/mainmatter.com: src/posts/2025-02-14-lttf-process.md

---
title: "Progressively Improving a Massive Codebase With Lint to the Future"
authorHandle: real_ate
tags: ["process"]
bio: "Senior Software Engineer, Creator of Lint to the Future"
description: "A walkthrough of the day-to-day process of using Lint to the Future to iteratively improve your codebase"
og:
  image:
tagline: |
  <p>
    With Lint to the Future you can quickly add new lint rules and track all the places that you still need to fix. In this article, we'll go into more detail about how to work through those big lists of files you have yet to fix all the lint rules for.
  </p>
---

## Intro

For anyone who has talked to me recently in a professional capacity, you will know that I'm low-key obsessed with using [Lint to the Future](https://github.com/mansona/lint-to-the-future) to gradually improve a codebase. This post isn't going to go much into what Lint to the Future is, but instead I'm going to showcase the day-to-day process I'm using to migrate ~6k files.

> ## TLDR;

Author: I've moved it to the bottom and expanded on what the point of it is :+1:

> I'll list the commands that you'll learn about in this post, feel free to keep reading to get the context:

Reviewer: the end of this sentence feels strange to me. I wonder if something along the lines of "in this post I'll be going through the relevance of these commands:" could be more suitable

Author: I'm going to move the TLDR; to the end as a summary of the commands that I introduced based on marco's comments :+1:

- `npx lttf ignore`
- `npx lttf remove my-custom-plugin/name-of-rule --filter 'app/**/*.[tj]s'`
- `eslint --no-eslintrc \`
  - `--parser-options "{ecmaVersion: 2021, sourceType: \"module\"}" \`
  - `--env "es6" --parser "@typescript-eslint/parser" \`
  - `--fix --no-inline-config --rule "{ my-custom-plugin/name-of-rule: error }" \`
  - `--plugin "my-custom-plugin" ./app/**/*.[tj]s`
- `npx codeowners-commit`

**Note:** That eslint command is all one command, and yes it's massive. Keep reading to understand why it needs to be so long.

> ## What is Lint to the Future

Reviewer 2: I'd make this section the start of the post before any of the above and change it from "Here's what I'm selling you" to "Here's a problem you might be facing"

Author: done

Like I said in the intro, I'm not going to go into this too much. [I've given a talk at EmberConf](https://youtu.be/Nl8gHDdkI0Y?si=3GOYECfPkGVsj_YW) if you want a bit more detail but the high level idea is this:

> You want to enable a new lint rule for your repo but you don't want to enable it and fix it for all the files at once. You know it's pointless enabling a lint rule as a warning because it will just get ignored and people will find it annoying. You want to draw a **temporary** line under the files that you currently have in the repo and make sure that any new files don't get introduced that violate the lint rule. And finally you want to be able to keep track of all the files that you temporarily ignored so you can come back and fix them over time.

Author: I was avoiding adding another list to this section, but you're right it's cleaner

This is the process that Lint to the Future enables 🎉 Once you have identified a new lint rule (or written a custom one) that you want to enable you follow these steps:

- enable that rule for your entire codebase as an **error**
- Install Lint to the Future (if you haven't already)
- Run `npx lttf ignore` to add file-based ignores to all the files that have lint errors
- Install the Lint to the Future dashboard to keep track of your files that you want to come back to

> I have used this process to improve a number of codebases in the past years and I have to say it feels glorious. I guess it's just a trick of our psycology that seeing the graph slowly go down over time can be a great source of dopamine.

Reviewer: ```suggestion
I have used this process to improve a number of codebases in the past years and I have to say it feels glorious. I guess it's just a trick of our psychology that seeing the graph slowly go down over time can be a great source of dopamine.
```

Author: done

> I have used this process to improve a number of codebases in the past years and I have to say it feels glorious. I guess it's just a trick of our psycology that seeing the graph slowly go down over time can be a great source of dopamine.

Reviewer 3: ```suggestion
I have used this process to improve a number of codebases in the past years and I have to say it feels glorious. I guess it's just a trick of our psychology that seeing the graph slowly go down over time can be a great source of dopamine.
```

Author: done

> But this just describes how to start using Lint to the Future, now that you have started using it what's the best way to work down those graphs?

Reviewer: ```suggestion
But this just describes how to start using Lint to the Future, now that you have started using it what's the best way to burn down those graphs?
```

I think repeating the last sentence as the next heading is more powerful here

Author: done

## Burning down the graphs

> As I said in the intro, I am currently working with a client that has a rule they want to fix across ~6k files. You might think, "fine, let's just run `eslint --fix` on the whole codebase and be done with it". I would not recommend this, even for lint rules that have a perfect fixer implementation with no risk of failure. The reason it's better to split the work is because of the **cognitive load** that it puts on the reviewer to give an approval to a massive PR. I talk about the briefly in my [git good blog post](/blog/2021/05/26/keeping-a-clean-git-history/), but you should have small targeted PRs that make a single logical change at a time.

Reviewer: This is the second time you said something along the lines of "as I said in the intro" and I don't think it's needed more than once

Author: I've removed the other one 👍

> As I said in the intro, I am currently working with a client that has a rule they want to fix across ~6k files. You might think, "fine, let's just run `eslint --fix` on the whole codebase and be done with it". I would not recommend this, even for lint rules that have a perfect fixer implementation with no risk of failure. The reason it's better to split the work is because of the **cognitive load** that it puts on the reviewer to give an approval to a massive PR. I talk about the briefly in my [git good blog post](/blog/2021/05/26/keeping-a-clean-git-history/), but you should have small targeted PRs that make a single logical change at a time.

Reviewer: ```suggestion
As I said in the intro, I am currently working with a client that has a rule they want to fix across ~6k files. You might think, "fine, let's just run `eslint --fix` on the whole codebase and be done with it". I would not recommend this, even for lint rules that have a perfect fixer implementation with no risk of failure. The reason it's better to split the work is because of the **cognitive load** that it puts on the reviewer to give an approval to a massive PR. I talk about this briefly in my [git good blog post](/blog/2021/05/26/keeping-a-clean-git-history/): you should have small, targeted PRs that make a single logical change at a time.
```

> As I said in the intro, I am currently working with a client that has a rule they want to fix across ~6k files. You might think, "fine, let's just run `eslint --fix` on the whole codebase and be done with it". I would not recommend this, even for lint rules that have a perfect fixer implementation with no risk of failure. The reason it's better to split the work is because of the **cognitive load** that it puts on the reviewer to give an approval to a massive PR. I talk about the briefly in my [git good blog post](/blog/2021/05/26/keeping-a-clean-git-history/), but you should have small targeted PRs that make a single logical change at a time.

Reviewer 3: I think it makes sense to change the ordering of the last 2 bits

```suggestion
As I said in the intro, I am currently working with a client that has a rule they want to fix across ~6k files. You might think, "fine, let's just run `eslint --fix` on the whole codebase and be done with it". I would not recommend this, even for lint rules that have a perfect fixer implementation with no risk of failure. The reason it's better to split the work is because of the **cognitive load** that it puts on the reviewer to give an approval to a massive PR. You should have small targeted PRs that make a single logical change at a time. I talk about the briefly in my [git good blog post](/blog/2021/05/26/keeping-a-clean-git-history/).
```

Author: done with a bit of rewording

> Here is the process that I'm currently using to burn down those 6k files

Reviewer: ```suggestion
Here is the process that I'm currently using to burn down those 6k files:
```

Author: done

### Remove all the lint ignores from a part of the app

> Lint to the future started with the ability to automatically add lint ignores to all your files. As people started to use it the very first feature request was to add the ability to remove lint ignores from all your files, this may seem like a silly request at first but it can be very tedious to remove ignoes from the top of each file you want to fix and using find and replace can be tempermental. So to remove a lint ignore across all your files you can run:

Reviewer: ```suggestion
Lint to the future started with the ability to automatically add lint ignores to all your files. As people started to use it, the very first feature request was to add the ability to remove lint ignores from all your files, this may seem like a silly request at first but it can be very tedious to remove ignores from the top of each file you want to fix and using find and replace can be temperamental. So to remove a lint ignore across all your files you can run:
```

Author: done

```
npx lttf remove my-custom-plugin/name-of-rule --filter 'app/**/*.[tj]s'
```

This will remove all of the file-based ignores from each of the JS or TS files in `app/`.

### Fix the rule if it has an auto fixer

> If you have a fast eslint setup this step is super easy

Reviewer: what do you mean by "fast" here?

Reviewer 2: I also don't understand – are you saying the below command would run for too long for 6k files? Even if that was the case and it ran 10min, what's the problem?

Reviewer 2: 🤯

Author: I've updated it to give more context

`eslint . --fix`

> But alas not all of us have a fast setup 😔 And you can probalby guess that if I'm working with such a large codebase that has 6k files to fix I don't have a fast eslint setup. Since you know which rule you want to target it would be pretty nice to run eslint with just one rule setup to improve the performance, but alas this isn't easy.

Reviewer 3: ```suggestion
But alas not all of us have a fast setup 😔 And you can probably guess that if I'm working with such a large codebase that has 6k files to fix I don't have a fast eslint setup. Since you know which rule you want to target it would be pretty nice to run eslint with just one rule setup to improve the performance, but alas this isn't easy.
```

> But alas not all of us have a fast setup 😔 And you can probalby guess that if I'm working with such a large codebase that has 6k files to fix I don't have a fast eslint setup. Since you know which rule you want to target it would be pretty nice to run eslint with just one rule setup to improve the performance, but alas this isn't easy.

Reviewer: ```suggestion
But alas not all of us have a fast setup 😔 And you can probably guess that if I'm working with such a large codebase that has 6k files to fix I don't have a fast eslint setup. Since you know which rule you want to target it would be pretty nice to run eslint with just one rule setup to improve the performance, but alas this isn't easy.
```

Author: fixed

If you search for "how to run eslint with one rule" you will get a number of different approachs, some plugins, and even an eslint wrapper that can do what you need. I tend to work with a carefully crafted command line execution that I edit accordingly. I'll show you the command first and then I'll break it down:

```
eslint --no-eslintrc --parser-options "{ecmaVersion: 2021, sourceType: \"module\"}" --env "es6" --parser "@typescript-eslint/parser" --fix --no-inline-config --rule "{ my-custom-plugin/name-of-rule: error }" --plugin "my-custom-plugin" ./app/**/*.[tj]s
```

> If you think this is a mouthful then you'd be right 🫠 It shouldn't really be this hard to execute eslint for a single rule but as I break it down you'll understand why it is.

Reviewer: ```suggestion
If you think this is a mouthful then you'd be right 🫠. It shouldn't really be this hard to execute eslint for a single rule but as I break it down you'll understand why it is.
```

Author: fixed

> The key thing here is that when we pass `--no-eslintrc` it's not going to load **anything** from your config, and that doesn't just mean rules. If you are using typescript in your project you probably have parser options that allow eslint to even read your files, this is what the first block of commands referrs to:

Reviewer: ```suggestion
The key thing here is that when we pass `--no-eslintrc` it's not going to load **anything** from your config, and that doesn't just mean rules. If you are using TypeScript in your project you probably have parser options that allow eslint to even read your files, this is what the first block of commands refers to:
```

Author: done

```
--parser-options "{ecmaVersion: 2021, sourceType: \"module\"}" --env "es6" --parser "@typescript-eslint/parser"
```

> Next we have `--fix` which is the same as the simple case. Then `--no-inline-config` will stop any extra config that you have added to files using elint comments from interfering with what we're trying to do.

Reviewer: ```suggestion
Next we have `--fix` which is the same as the simple case. Then `--no-inline-config` will stop any extra config that you have added to files using eslint comments from interfering with what we're trying to do.
```

Then we need to setup your custom rule:

```
--rule "{ my-custom-plugin/name-of-rule: error }" --plugin "my-custom-plugin"
```

If you've written your own eslint rule in your own plugin it's not enough to just pass the rule name, you need to tell eslint about which plugin that rule comes from.

> > Bonus argument: If you're working in a monorepo and want to run things from the root but there are eslint ignores stopping you then you can add `--no-ignore` to the command. ⚠️ this can cause to you run the fix on files that you actually didn't intend to so be careful.

Reviewer 3: ```suggestion
> Bonus argument: If you're working in a monorepo and want to run things from the root but there are eslint ignores stopping you then you can add `--no-ignore` to the command. ⚠️ This can cause to you run the fix on files that you actually didn't intend to so be careful.
```

Author: done

> > Bonus argument: If you're working in a monorepo and want to run things from the root but there are eslint ignores stopping you then you can add `--no-ignore` to the command. ⚠️ this can cause to you run the fix on files that you actually didn't intend to so be careful.

Author: fixed

- grouped by their codeowner and
- into batches with a maximum of 100 files (by default)

If you don't have a codeowners file that's fine, that would just be a empty codeowner for each file and you would only be batching by maximum files in this case. To run it all you need to do is:

```
npx codeowners-commit
```

And your current branch will have a number of new batched commits that you can cherry-pick into new PRs to your heart's content!

> > Bonus argument: if the default of 100 files per commit is just too little then you can pass `--max-files 200` and it will bump up the max files per batch 👍

Reviewer: ```suggestion
> Bonus argument: if the default of 100 files per commit is just too little then you can pass `--max-files <number_of_files>` and it will bump up the max files per batch 👍
```
unless 200 is the only valid option

Reviewer 2: ```suggestion
Bonus argument: if the default of 100 files per commit is just too little then you can pass `--max-files 200` and it will bump up the max files per batch 👍
```

This isn't a quote. I understand everyone seems to be missing some kind of callout/note element and we should maybe add that but in absence of one, we shouldn't misuse quotes.

## Conclusion

> If you found this helpful please [let me know on Bluesky](https://bsky.app/profile/chris.manson.ie) and feel free to send me your awesome graphs going down and to the right, together we can make our code better one step at a time 💪

Reviewer: ```suggestion
If you found this helpful please [let me know on Bluesky](https://bsky.app/profile/chris.manson.ie) and feel free to send me your awesome graphs going down and to the right, together we can make our code better, one step at a time 💪
```

Author: fixed

---
title: "Progressively Improving a Massive Codebase With Lint to the Future"
authorHandle: real_ate
tags: ["process"]
bio: "Senior Software Engineer, Creator of Lint to the Future"
description: "A walkthrough of the day-to-day process of using Lint to the Future to iteratively improve your codebase"
og:
  image:
tagline: |
  <p>

>     With Lint to the Future you can quickly add new lint rules and track all the places you still need to fix. In this article, we'll go into more detail about how to work through those big lists of files you have yet to fix all the lint rules for.

Reviewer 2: I'd again strongly recommend to start with describing what the problem is that lttf fixes. I don't really think it's

> With Lint to the Future you can quickly add new lint rules and track all the places you still need to fix.

but actually the problem is that if you enable a new lint rule in a massive codebase, you need to update the code in a thousand places in a huge PR? Or you add the rule as a warning which makes it pointless. Either way, the problem is there's no real way to introduce new rules step-by-step which is the actual problem that lttf solves (even with a burndown chart ootb)?

Reviewer 2: I really think this should be added in 1-2 sentences in the beginning since otherwise, nobody who doesn't already know what lttf is, will be interested in this… I see that's coming in more detail below but you need to have a super short version of this in the beginning or you won't get people to read far enough to see the more detailed explanation below

Author: Ah so actually there was a bit of a communication breakdown. When you were saying "intro" you were actually specifically talking about the **tagline**. That's why all my previous comments were referring to improving the first two paragraphs of the body of the post rather than the tagline 🙈 

I'll improve the tagline now 👍

Reviewer 2: I guess this attributed is named misleadingly – it's rendered as the beginning of the blog post

  </p>
---

## Intro

> Anyone who has talked to me recently in a professional capacity will know that I'm low-key obsessed with using [Lint to the Future](https://github.com/mansona/lint-to-the-future) to improve a codebase gradually. This post won't go much into what Lint to the Future is, but instead, I'm going to showcase the day-to-day process I'm using to migrate ~6k files.

Reviewer 2: ```suggestion
Anyone who has talked to me recently in a professional capacity will know that I'm low-key obsessed with using [Lint to the Future](https://github.com/mansona/lint-to-the-future) to improve a codebase gradually. I'm going to showcase the day-to-day process I'm using to migrate ~6k files.
```

Right after this comes the paragraph that explains what lttf is

- You want to enable a new lint rule for your repo but you don't want to enable it and fix it for all the files all at once.
- You know it's pointless enabling a lint rule as a warning because it will just get ignored and people will find it annoying.
- You want to draw a **temporary** line under the files that you currently have in the repo
- You want to ensure that nobody introduces **new** files violating the lint rule to the codebase.
- Finally, you want to be able to keep track of all the files that you temporarily ignored so you can come back and fix them over time.

This is the situation that Lint to the Future is designed to help you with. Once you have identified a new lint rule (or written a custom one) that you want to enable you follow these steps:

- Enable that rule for your entire codebase as an **error**
- Install Lint to the Future (if you haven't already)
- Run `npx lttf ignore` to add file-based ignores to all the files that have lint errors
- Install the Lint to the Future dashboard to keep track of the files that you want to come back to

> 

Reviewer 2: a screenshot of an example graph would be really good here

Author: added 👍

Here is the process that I'm currently using to burn down those 6k files:

### Remove all the lint ignores from a part of the app

I have used this process to improve a number of codebases in the past years and I have to say it feels glorious. I guess it's just a trick of our psychology that seeing the graph slowly go down over time can be a great source of dopamine. Lint to the future started with the ability to automatically add lint ignores to all your files. As people started to use it the very first feature request was to add the ability to remove lint ignores from all your files, this may seem like a silly request at first but it can be very tedious to remove ignoes from the top of each file you want to fix and using find and replace can be temperamental. So to remove a lint ignore across all your files you can run:

```
npx lttf remove my-custom-plugin/name-of-rule --filter 'app/**/*.[tj]s'
```

This will remove all of the file-based ignores from each of the JS or TS files in `app/`.

### Fix the rule if it has an auto fixer

> For most repos running eslint should be super quick so you can fix rules using the normal recommended eslint command:

Reviewer: ```suggestion
For most repos, running eslint should be super quick so you can fix rules using the normal recommended eslint command:
```

`eslint . --fix`

> Unfortunately, not all of us have a fast setup 😔 Because I'm working with such a large codebase that has 6k files to fix, my eslint is a bit slower than usual, but the real problem that we face with this repo is the use of the typescript-eslint plugin. With typescript-eslint it has to do a complete `tsc` compilation of your codebase before it can give you any type information for your lint rules, and if your `tsc` is slow your `eslint` is doubly slow. Since you know which rule you want to target it would be nice to run eslint with just one rule setup to improve the performance, but alas this isn't easy.

Reviewer 2: ```suggestion
Unfortunately, not all of us have a fast setup 😔 Because I'm working with such a large codebase that has 6k files to fix, my eslint is a bit slower than usual, but the real problem that we face with this repo is the use of the [`typescript-eslint` plugin](https://typescript-eslint.io). With `typescript-eslint` it has to do a complete `tsc` compilation of your codebase before it can give you any type information for your lint rules, and if your `tsc` is slow your `eslint` is doubly slow. Since you know which rule you want to target it would be nice to run `eslint` with just one rule setup to improve the performance, but alas this isn't easy.
```

> But alas not all of us have a fast setup 😔 And you can probably guess that if I'm working with such a large codebase that has 6k files to fix my eslint is a bit slower than usual, but the real problem that we face with this repo is the use of the tyepscript-eslint plugin. With typescript-eslint it has to do a full `tsc` compilation of your codebase before it can give you any type information for your lint rules, and if your `tsc` is slow your `eslint` is doubly slow. Since you know which rule you want to target it would be pretty nice to run eslint with just one rule setup to improve the performance, but alas this isn't easy.

Reviewer: ```suggestion
But alas not all of us have a fast setup 😔 And you can probably guess that if I'm working with such a large codebase that has 6k files to fix my eslint is a bit slower than usual, but the real problem that we face with this repo is the use of the typescript-eslint plugin. With typescript-eslint it has to do a full `tsc` compilation of your codebase before it can give you any type information for your lint rules, and if your `tsc` is slow, your `eslint` is doubly slow. Since you know which rule you want to target it would be pretty nice to run eslint with just one rule set up to improve the performance, but alas this isn't easy.
```

"any type information" is this regarding typing, or should this be "any type of information"?

This paragraph has two "but alas"s in it.

"And you can probably guess that if I'm working with such a large codebase that has 6k files to fix my eslint is a bit slower than usual" doesn't feel like a very natural sentence, unfortunately I'm blanking on a suggestion for it though.

> If you search for "how to run eslint with one rule" you will get a number of different approachs, some plugins, and even an eslint wrapper that can do what you need. I tend to work with a carefully crafted command line execution that I edit accordingly. I'll show you the command first and then I'll break it down:

Reviewer: ```suggestion
If you search for "how to run eslint with one rule" you will get a number of different approaches, some plugins, and even an eslint wrapper that can do what you need. I tend to work with a carefully crafted command line execution that I edit accordingly. I'll show you the command first and then break it down:
```

```
--parser-options "{ecmaVersion: 2021, sourceType: \"module\"}" --env "es6" --parser "@typescript-eslint/parser"
```

Next we have `--fix` which is the same as the simple case. Then `--no-inline-config` will stop any extra config that you have added to files using eslint comments from interfering with what we're trying to do.

Then we need to setup your custom rule:

```
--rule "{ my-custom-plugin/name-of-rule: error }" --plugin "my-custom-plugin"
```

If you've written your own eslint rule in your own plugin it's not enough to just pass the rule name, you need to tell eslint about which plugin that rule comes from.

**Bonus argument:** If you're working in a monorepo and want to run things from the root but there are eslint ignores stopping you then you can add `--no-ignore` to the command. ⚠️ This can cause to you run the fix on files that you actually didn't intend to so be careful.

### Batch commit your fixes

> If you're following along with the steps you should have a whole bunch of files in your app folder changed but not committed to git yet. If we're going to split these up to separate PRs we need a way to easily split these commits, and I have a tool for that! I created the tool [codeowners-commit](https://github.com/mansona/codeowners-commit) for exactly this purpose. If you're working on a project that has 6k files to fix you will likely also already know about a [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) file. You might even have some sort of review process that requires a team to review their code. `codeowners-commit` batches files into commits that are

Reviewer: ```suggestion
If you're following along with the steps you should have a whole bunch of files in your app folder changed but not committed to git yet. If we're going to split these up to separate PRs we need a way to easily split these commits, and I have a tool for that! I created the tool [codeowners-commit](https://github.com/mansona/codeowners-commit) for exactly this purpose. If you're working on a project that has 6k files to fix you will likely also already know about a [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) file. You might even have some sort of review process that requires a team to review their code. `codeowners-commit` batches files into commits that are:
```

> - grouped by their codeowner and
> - into batches with a maximum of 100 files (by default)

Reviewer: this doesn't really make sense as bullet points as it is.
```suggestion
- grouped by their codeowner
- grouped into batches with a maximum of 100 files (by default)
```

---
title: "Progressively Improving a Massive Codebase With Lint to the Future"
authorHandle: real_ate
tags: ["process"]
bio: "Senior Software Engineer, Creator of Lint to the Future"
description: "A walkthrough of the day-to-day process of using Lint to the Future to iteratively improve your codebase"
og:
  image: "/assets/images/posts/2025-02-14-lttf-process/lttf-process-og.jpg"
tagline: |
  <p>

>     When was the last time that you wanted to add a new lint rule to a massive codebase but decided against it becuase it would be too much effort to fix all the existing files? Maybe you though about turning the new rule on as a warning but realised it was pointless because lint-warnings are too easy to ignore. This post offers a better way to add new lint rules to existing codebases with Lint to the Future, and teaches you the best way to progressively improve your codebase over time.

Reviewer 4: ```suggestion
    When was the last time that you wanted to add a new lint rule to a massive codebase but decided against it because it would be too much effort to fix all the existing files? Maybe you though about turning the new rule on as a warning but realised it was pointless because lint-warnings are too easy to ignore. This post offers a better way to add new lint rules to existing codebases with Lint to the Future, and teaches you the best way to progressively improve your codebase over time.
```
